File: g01-android-game/app/src/main/java/sdk/vlplay/vn/sample/design/components/VLTopAppBar.kt
package sdk.vlplay.vn.sample.design.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.DeveloperMode
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import sdk.vlplay.vn.sample.design.theme.VLColor
import sdk.vlplay.vn.sample.design.theme.VLRadius
import sdk.vlplay.vn.sample.design.theme.VLSpacing
import sdk.vlplay.vn.sample.design.theme.VLTheme

/**
 * Top app bar — 2 variants per mockup. Parity iOS [VLTopAppBar].
 *
 * - [Debugger] "SDK_DEBUGGER" + developer-mode icon (root tab screens, gray tone)
 * - [Brand] "VLPlay SDK" + red diamond logo (detail screens)
 */
enum class VLTopBarVariant { Debugger, Brand }

@Composable
fun VLTopAppBar(
    modifier: Modifier = Modifier,
    variant: VLTopBarVariant = VLTopBarVariant.Debugger,
    subtitle: String? = null,
    environmentBadge: String? = null,
    trailing: @Composable () -> Unit = {},
) {
    // Layout parity iOS VLTopAppBar.swift: logo · title/subtitle · env-pill · Spacer · trailing.
    // The Column wraps its content (no weight); a trailing Spacer pushes `trailing()` to the
    // far right so the env-pill sits adjacent to the title instead of stretched apart.
    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(VLColor.Surface)
            .padding(horizontal = VLSpacing.safeMargin, vertical = VLSpacing.md),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(VLSpacing.sm),
    ) {
        LogoMark(variant = variant)
        Column {
            Text(
                text = if (variant == VLTopBarVariant.Debugger) "SDK_DEBUGGER" else "VLPlay SDK",
                style = MaterialTheme.typography.titleLarge.copy(
                    fontWeight = FontWeight.SemiBold,
                ),
                color = if (variant == VLTopBarVariant.Debugger) VLColor.OnSurfaceVariant else VLColor.Primary,
                overflow = TextOverflow.Ellipsis,
            )
            subtitle?.let {
                Text(
                    text = it,
                    style = MaterialTheme.typography.bodySmall,
                    color = VLColor.OnSurfaceVariant,
                )
            }
        }
        environmentBadge?.let {
            VLStatePill(
                text = it,
                tone = if (variant == VLTopBarVariant.Debugger) VLPillTone.Neutral else VLPillTone.Primary,
            )
        }
        Spacer(modifier = Modifier.weight(1f))
        trailing()
    }
}

@Composable
private fun LogoMark(variant: VLTopBarVariant) {
    when (variant) {
        VLTopBarVariant.Debugger -> Box(
            modifier = Modifier
                .size(28.dp)
                .clip(RoundedCornerShape(VLRadius.sm))
                .background(VLColor.SurfaceContainerHigh),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = Icons.Filled.DeveloperMode,
                contentDescription = null,
                tint = VLColor.OnSurfaceVariant,
                modifier = Modifier.size(14.dp),
            )
        }
        VLTopBarVariant.Brand -> Box(
            modifier = Modifier.size(32.dp),
            contentAlignment = Alignment.Center,
        ) {
            Box(
                modifier = Modifier
                    .size(28.dp)
                    .rotate(45f)
                    .clip(RoundedCornerShape(VLRadius.sm))
                    .background(VLColor.Primary),
            )
            Text(
                text = "VL",
                style = MaterialTheme.typography.labelSmall.copy(
                    fontWeight = FontWeight.ExtraBold,
                    fontSize = 11.sp,
                ),
                color = Color.White,
            )
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun VLTopAppBarDebuggerPreview() {
    VLTheme {
        VLTopAppBar(
            variant = VLTopBarVariant.Debugger,
            subtitle = "Phase 0 ready",
            environmentBadge = "STAGING",
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun VLTopAppBarBrandPreview() {
    VLTheme {
        VLTopAppBar(
            variant = VLTopBarVariant.Brand,
            subtitle = "Profile",
            environmentBadge = "PROD",
        )
    }
}

File: g01-ios-game/VLPlayDemo2/Design/VLTopAppBar.swift
import SwiftUI

// Top app bar — 2 variants per mockup:
//   .debugger : "SDK_DEBUGGER" + developer-mode icon (root tab screens, gray tone)
//   .brand    : "VLPlay SDK" + red diamond logo (detail screens)

enum VLTopBarVariant {
    case debugger
    case brand
}

struct VLTopAppBar<Trailing: View>: View {
    var variant: VLTopBarVariant = .debugger
    var subtitle: String? = nil
    var environmentBadge: String? = nil  // STAGING / PROD pill
    @ViewBuilder var trailing: () -> Trailing

    init(variant: VLTopBarVariant = .debugger,
         subtitle: String? = nil,
         environmentBadge: String? = nil,
         @ViewBuilder trailing: @escaping () -> Trailing = { EmptyView() }) {
        self.variant = variant
        self.subtitle = subtitle
        self.environmentBadge = environmentBadge
        self.trailing = trailing
    }

    var body: some View {
        HStack(spacing: VLSpacing.sm) {
            logoMark
            VStack(alignment: .leading, spacing: 0) {
                Text(titleText)
                    .font(VLFont.h2)
                    .foregroundColor(titleColor)
                    .tracking(variant == .debugger ? 0.5 : 0)
                if let subtitle {
                    Text(subtitle)
                        .font(VLFont.bodySm)
                        .foregroundColor(VLColor.onSurfaceVariant)
                }
            }
            if let badge = environmentBadge {
                VLStatePill(text: badge, tone: variant == .debugger ? .neutral : .primary)
            }
            Spacer()
            trailing()
        }
        .padding(.horizontal, VLSpacing.safeMargin)
        .padding(.vertical, VLSpacing.md)
        .background(VLColor.surface)
    }

    private var titleText: String {
        switch variant {
        case .debugger: return "SDK_DEBUGGER"
        case .brand:    return "VLPlay SDK"
        }
    }

    private var titleColor: Color {
        switch variant {
        case .debugger: return VLColor.onSurfaceVariant
        case .brand:    return VLColor.primary
        }
    }

    @ViewBuilder
    private var logoMark: some View {
        switch variant {
        case .debugger:
            ZStack {
                RoundedRectangle(cornerRadius: VLRadius.sm, style: .continuous)
                    .fill(VLColor.surfaceContainerHigh)
                    .frame(width: 28, height: 28)
                Image(systemName: VLIcon.developerMode)
                    .font(.system(size: 14, weight: .bold))
                    .foregroundColor(VLColor.onSurfaceVariant)
            }
        case .brand:
            ZStack {
                RoundedRectangle(cornerRadius: VLRadius.sm, style: .continuous)
                    .fill(VLColor.primary)
                    .frame(width: 28, height: 28)
                    .rotationEffect(.degrees(45))
                Text("VL")
                    .font(.system(size: 11, weight: .heavy))
                    .foregroundColor(.white)
            }
            .frame(width: 32, height: 32)
        }
    }
}

extension VLTopAppBar where Trailing == EmptyView {
    init(variant: VLTopBarVariant = .debugger,
         subtitle: String? = nil,
         environmentBadge: String? = nil) {
        self.init(variant: variant,
                  subtitle: subtitle,
                  environmentBadge: environmentBadge,
                  trailing: { EmptyView() })
    }
}
